### PR TITLE
Added ResonanceMolSupplier

### PR DESCRIPTION
**Pull request by ptosco**

### CODE REVIEW

File: Code/GraphMol/Resonance.cpp [hunk 1]
```diff
@@ -0,0 +1,1355 @@
...
+//  This file is part of the RDKit.
+//  The contents are covered by the terms of the BSD license
+//  which is included in the file license.txt, found at the root
+//  of the RDKit source tree.
+//
+#include <iostream>
+#include <GraphMol/RDKitBase.h>
+#include <GraphMol/MolOps.h>
+#include "Resonance.h"
+#include <boost/functional/hash.hpp>
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: I'm not sure if it's still true, but the boost hashing code used to generate different results on 32 and 64 bit machines. It's probably not important here that you get consistent hashes across architectures, but if it is, you should probably switch to using the code in $RDBASE/Code/RDGeneral/hash (derived from an older version of boost.hash).

File: Code/GraphMol/Resonance.cpp [hunk 2]
```diff
@@ -0,0 +1,1355 @@
...
+            break;
+          case MolOps::SANITIZE_SETAROMATICITY:
+            s = "SANITIZE_SETAROMATICITY";
+            break;
+          case MolOps::SANITIZE_ADJUSTHS:
+            s = "SANITIZE_ADJUSTHS";
+            break;
+        }
+        s += " failed";
+        throw std::runtime_error(s);
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: I'm not sure I understand the logic here. Why not just let the underlying sanitization code throw an exception?

File: Code/GraphMol/Resonance.cpp [hunk 3]
```diff
@@ -0,0 +1,1355 @@
...
+    d_atom(ae.d_atom),
+    d_parent(parent)
+  {
+  }
+  
+  // assign total valence, formal charge and non-bonded electrons
+  // from the original atom
+  void AtomElectrons::initTvNbFcFromAtom()
+  {
+    d_tv = d_atom->getTotalValence();
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: I think this probably needs a PRECONDITION that d_atom is not NULL

File: Code/GraphMol/Resonance.cpp [hunk 4]
```diff
@@ -0,0 +1,1355 @@
...
+    ROMol::OEDGE_ITER nbrIdx, endNbrs;
+    boost::tie(nbrIdx, endNbrs) = mol.getAtomBonds(d_atom);
+    for (; !res && (nbrIdx != endNbrs); ++nbrIdx) {
+      const Bond *bondNbr = mol[*nbrIdx].get();
+      unsigned int biNbr = bondNbr->getIdx();
+      if (d_parent->parent()->getBondConjGrpIdx(biNbr) != conjGrpIdx())
+        continue;
+      BondElectrons *beNbr =
+        d_parent->getBondElectronsWithIdx(biNbr);
+      const Atom *atomNbr = ((bondNbr->getBeginAtom() == d_atom)
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: This is more simply done with Bond::getOtherAtom()

File: Code/GraphMol/Resonance.cpp [hunk 5]
```diff
@@ -0,0 +1,1355 @@
...
+    boost::uint8_t canAdd = !isDefinitive();
+    if (canAdd)
+      canAdd = (d_tv <= (5 - bo));
+    // if canAdd is true, loop over neighboring conjugated bonds
+    // and check their definitive flag; if all neighbors are
+    // definitive, we need an additional check on total valence
+    if (canAdd && isLastBond()) {
+      // we allow a formal charge up to 2 on atoms right of
+      // carbon, not more than 1 on atoms left of N
+      unsigned int rightOfC = ((d_atom->getAtomicNum() > 6) ? 1 : 0);
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: I haven't found a counter-example yet, but us it this going to work for atoms not in the first row of the periodic table like Si?

File: Code/GraphMol/Resonance.cpp [hunk 6]
```diff
@@ -0,0 +1,1355 @@
...
+      unsigned int nbi = mol[*nbrIdx].get()->getIdx();
+      if ((nbi != bi)
+        && (d_parent->parent()->getBondConjGrpIdx(nbi) == conjGrpIdx()))
+        allDefinitive =
+          d_parent->getBondElectronsWithIdx(nbi)->isDefinitive();
+    }
+    if (allDefinitive)
+      setLastBond();
+  }
+
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: IGNORE: marker to show how far I've made it through.

File: Code/GraphMol/Resonance.cpp (hunk 4 above)
Comment (ptosco) [inline Code/GraphMol/Resonance.cpp]: Hi Greg,

Thanks for your feedback, I'll make those changes and check the hashing behaviour in Boost.

Cheers,
p.

File: Code/GraphMol/Resonance.cpp [hunk 7]
```diff
@@ -0,0 +1,1355 @@
...
+      it != ce.d_conjBondMap.end(); ++it)
+      d_conjBondMap[it->first] = new BondElectrons(this, *(it->second));
+  }
+
+  // object destructor
+  ConjElectrons::~ConjElectrons()
+  {
+    for (ConjAtomMap::const_iterator it = d_conjAtomMap.begin();
+      it != d_conjAtomMap.end(); ++it) {
+      if (it->second)
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: This test isn't required.

File: Code/GraphMol/Resonance.cpp [hunk 8]
```diff
@@ -0,0 +1,1355 @@
...
+        delete it->second;
+    }
+    for (ConjBondMap::const_iterator it = d_conjBondMap.begin();
+      it != d_conjBondMap.end(); ++it) {
+      if (it->second)
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: This test isn't required.

File: Code/GraphMol/Resonance.cpp [hunk 9]
```diff
@@ -0,0 +1,1355 @@
...
+  void ConjElectrons::assignFormalCharge()
+  {
+    for (ConjAtomMap::const_iterator it = d_conjAtomMap.begin();
+      it != d_conjAtomMap.end(); ++it)
+      it->second->assignFormalCharge();
+  }
+  
+  // return true if formal charges for this ConjElectrons
+  // object are OK, false if they aren't
+  bool ConjElectrons::checkCharges()
```
Comment (greglandrum) [inline Code/GraphMol/Resonance.cpp]: IGNORE: placeholder for how far I've read

**Pull request merged**